cmd/sonobuoy/app/runner.go

```diff
@@ -0,0 +1,330 @@
+	runnerset := pflag.NewFlagSet("runner", pflag.ExitOnError)
+	// Default to detect since we need kubeconfig regardless
+	runnerset.AddFlagSet(GenFlagSet(&cfg.genFlags, DetectRBACMode, ConformanceImageVersionAuto))
+	AddSkipPreflightFlag(&cfg.skipPreflight, runnerset)
+
+	runnerset.DurationVar(
+		&cfg.timeout, "runner-timeout", 6*time.Hour,
+		"Length of time to give the runner before giving up. (Default: 6h)",
+	)
+	runnerset.DurationVar(
+		&cfg.pollingInterval, "polling-interval", 5*time.Minute,
+		"Duration of time between polling sonobuoy for status. (Default: 5m)",
+	)
+	runnerset.DurationVar(
+		&cfg.statusLoopDelay, "status-loop-delay", 1*time.Minute,
+		"Duration of time between starting tests and checking the status. (Default: 1m)",
+	)
+	runnerset.DurationVar(
+		&cfg.retrievalDelay, "retrieval-delay", 15*time.Second,
+		"Duration of time between getting completed status and retrieving results. (Default: 15s)",
+	)
+	runnerset.BoolVar(
+		&cfg.showAll, "show-all", false,
+		"Don't summarize plugin statuses, show all individually",
+	)
+
+	return runnerset
+}
+
+func (r *runnerFlags) Config() (*client.RunConfig, error) {
+	gencfg, err := r.genFlags.Config()
+	if err != nil {
+		return nil, err
+	}
+	return &client.RunConfig{
+		GenConfig: *gencfg,
+	}, nil
+}
+
+func init() {
+	cmd := &cobra.Command{
+		Use:   "runner",
+		Short: "Submits a sonobuoy run and manages the lifetime of that run",
+		Run:   startSonobuoyRunner,
+		Args:  cobra.ExactArgs(0),
+	}
+
+	cmd.Flags().AddFlagSet(RunnerFlagSet(&runnerflags))
+	RootCmd.AddCommand(cmd)
+}
+
+func startSonobuoyRunner(cmd *cobra.Command, args []string) {
+	outDir := defaultOutDir
+	if len(args) > 0 {
+		outDir = args[0]
+	}
+
+	var signals chan os.Signal
+	signals = make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM, syscall.SIGUSR1)
+
+	cfg, err := runnerflags.kubecfg.Get()
+	if err != nil {
+		errlog.LogError(errors.Wrap(err, "couldn't get REST client"))
+		os.Exit(1)
+	}
+
+	runCfg, err := runnerflags.Config()
+	if err != nil {
+		errlog.LogError(errors.Wrap(err, "could not retrieve E2E config"))
+		os.Exit(1)
+	}
+	sonobuoyClient, err = getSonobuoyClient(cfg)
+	if err != nil {
+		errlog.LogError(errors.Wrap(err, "could not create sonobuoy client"))
+		os.Exit(1)
+	}
+
+	failed := run(runCfg)
+
+	if failed {
+		os.Exit(1)
+	}
+
+	go func() {
+		_, open := <-signals
+		delete()
+
+		if open {
+			close(signals)
+			os.Exit(1)
+		} else {
+			os.Exit(0)
+		}
+	}()
+
+	time.Sleep(runnerflags.statusLoopDelay)
+
+	timeout := time.Now().Add(runnerflags.timeout)
+	for {
+		delta := time.Now()
+
+		if delta.After(timeout) {
+			errlog.LogError(errors.New("Runner timed out waiting for tests to complete"))
+			signals <- syscall.SIGUSR1
+			return
+		}
+
+		status, err := status()
+		if err != nil {
+			errlog.LogError(err)
+			continue
+		}
+
+		switch status.Status {
+		case aggregation.FailedStatus:
+			signals <- syscall.SIGUSR1
```

Comment:
I don't love this. I get that passing in SIGUSR1 makes a lot of sense for the end user, but using it internally feels like a hack.